include/boost/geometry/algorithms/detail/buffer/piece_border.hpp

```diff
@@ -0,0 +1,509 @@
+    std::size_t m_end;
+
+    // Points from the original (one or two, depending on piece shape)
+    // Note, if there are 2 points, they are REVERSED w.r.t. the original
+    // Therefore here we can walk in its order.
+    boost::array<Point, 2> m_originals;
+    std::size_t m_original_size;
+
+    geometry::model::box<Point> m_envelope;
+    bool m_has_envelope;
+
+    // True if piece is determined as "convex"
+    bool m_is_convex;
+
+    // True if offsetted part is monotonically changing in x-direction
+    bool m_is_monotonic_increasing;
+    bool m_is_monotonic_decreasing;
+
+    radius_type m_min_comparable_radius;
+    radius_type m_max_comparable_radius;
+
+    piece_border()
+        : m_reversed(false)
+        , m_ring(NULL)
+        , m_begin(0)
+        , m_end(0)
+        , m_original_size(0)
+        , m_has_envelope(false)
+        , m_is_convex(false)
+        , m_is_monotonic_increasing(false)
+        , m_is_monotonic_decreasing(false)
+        , m_min_comparable_radius(0)
+        , m_max_comparable_radius(0)
+    {
+    }
+
+    //! Returns true if there is a ring, the ring has one or more points,
+    //! and the original has one or two points
+    bool valid() const
+    {
+        return m_ring != NULL && m_begin < m_end && m_original_size > 0;
+    }
+
+    // Only used for debugging (SVG)
+    Ring get_full_ring() const
+    {
+        Ring result;
+        if (! valid())
+        {
+            return result;
+        }
+        std::copy(m_ring->begin() + m_begin,
+                  m_ring->begin() + m_end,
+                  std::back_inserter(result));
+        std::copy(m_originals.begin(),
+                  m_originals.begin() + m_original_size,
+                  std::back_inserter(result));
+        // Add the closing point
+        result.push_back(*(m_ring->begin() + m_begin));
+
+        return result;
+    }
+
+    void get_properties_of_border(bool is_point_buffer, Point const& center)
+    {
+        m_has_envelope = calculate_envelope(m_envelope);
+        if (m_has_envelope)
+        {
+            // Take roundings into account, enlarge box
+            detail::buffer::buffer_box(m_envelope, 1.0e-3, m_envelope);
+        }
+        if (valid() && is_point_buffer)
+        {
+            // Determine min/max radius
+            calculate_radii(center, m_ring->begin() + m_begin, m_ring->begin() + m_end);
+        }
+    }
+
+    template <typename SideStrategy>
+    void get_properties_of_offsetted_ring_part(SideStrategy const& strategy)
+    {
+        if (valid())
+        {
+            m_is_convex = is_convex(strategy);
+            check_monotonicity(m_ring->begin() + m_begin, m_ring->begin() + m_end);
+        }
+    }
+
+    void set_offsetted(Ring const& ring, std::size_t begin, std::size_t end)
+    {
+        BOOST_GEOMETRY_ASSERT(begin <= end);
+        BOOST_GEOMETRY_ASSERT(begin < boost::size(ring));
+        BOOST_GEOMETRY_ASSERT(end <= boost::size(ring));
+
+        m_ring = &ring;
```

Comment:
Can `Ring` be a user-defined type or is it our type and we can be sure that the `operator&` is not defined? Because otherwise `boost::addressof` should be used here.

Reply:
No, the ring is an offsetted ring and defined by ourselves.